app/ui/client/views/app/components/hooks.js

```diff
@@ -0,0 +1,67 @@
+import { useState, useEffect, useMemo, useCallback } from 'react';
+import moment from 'moment';
+
+import { useUserPreference } from '../../../../../../client/contexts/UserContext';
+import { useSetting } from '../../../../../../client/contexts/SettingsContext';
+
+export function useDebounce(value, delay) {
+	const [debouncedValue, setDebouncedValue] = useState(value);
+
+	useEffect(() => {
+		const handler = setTimeout(() => {
+			setDebouncedValue(value);
+		}, delay);
+		return () => clearTimeout(handler);
+	}, [value]);
+
+	return debouncedValue;
+}
+
+export function useQuery(params, sort, type, workspace = 'local') {
+	return useMemo(() => ({
+		query: JSON.stringify({
+			type,
+			text: params.term,
+			workspace,
+		}),
+		sort: JSON.stringify({ [sort[0]]: sort[1] === 'asc' ? 1 : 0 }),
+		...params.itemsPerPage && { count: params.itemsPerPage },
+		...params.current && { offset: params.current },
+	}), [params, sort, type, workspace]);
+}
+
+const dayFormat = ['h:mm A', 'H:mm'];
+export function useFormatTime() {
+	const clockMode = useUserPreference('clockMode', false);
+	const sameDay = dayFormat[clockMode - 1] || useSetting('Message_TimeFormat');
```

Comment:
The `useSetting()` hook is conditionally called here, which violates the Rules of Hooks and may trigger errors.